Shared receiver can start serving before consumers of its   other-signal pipelines are started

### Component(s)

service

### What happened?

**Describe the bug**

If you share one OTLP receiver across pipelines of different signals (say one `otlp` receiver feeding traces, logs and metrics), the order in which components start is not deterministic between those pipelines.
And since the OTLP receiver is a sharedcomponent, whichever per-signal receiver node happens to start first brings up the single HTTP/gRPC server for all signals at once.
So the server can start accepting data for a signal whose pipeline hasn't finished starting yet.

I'm fairly sure this is what's actually behind the k8sattributes startup panic in [contrib #47672][contrib-47672] (`processor.go:202`).
k8sattributes only sets its kube client (`kc`, an interface) in `Start()`.
If a span arrives before that `Start()` has run, `kc` is still nil and the call dispatches on a nil interface.

The usual "reverse-topological start guarantees consumers are up before the receiver" reasoning holds, but only inside a single connected pipeline.
It doesn't carry across signals for a shared receiver, because the graph doesn't treat that receiver as one node.

Here's the chain as I read it:

- A shared receiver gets one node *per signal*, not one node total. The node ID comes from `(signal, componentID)` (see `newReceiverNode` in `service/internal/graph/receiver.go`). With a different exporter per signal, the per-signal subgraphs end up disconnected from each other:

  ```
  recv[traces] -> proc[traces] -> exp[traces]
  recv[logs]   -> proc[logs]   -> exp[logs]
  recv[metrics]-> proc[metrics]-> exp[metrics]
  ```

- `StartAll` does `topo.Sort(componentGraph)` and starts in reverse order (`service/internal/graph/graph.go`). gonum keeps the nodes in a map, so the sort doesn't fix any relative order *between* the disconnected subgraphs. There's no barrier or two-phase "start everything, then serve".

- The single server registers handlers for every signal at build time, not at start time (otlpreceiver's `startHTTPServer` wires `/v1/traces`, `/v1/metrics`, `/v1/logs` together). So the first receiver node to start opens the door for all three.

Put together: reverse-topo does guarantee `recv[traces]` starts after the traces processors, but the server isn't tied to `recv[traces]` — it's brought up by whichever signal's node wins the race.
If that's `recv[logs]` and the traces pipeline's k8sattributes `Start()` hasn't run, a request comes in, gets routed into a chain whose `kc` is still nil, and you hit `processor.go:202`.

Worth noting it's not specific to traces.
`processResource` is shared by processTraces/Logs/Metrics/Profiles, so a cluster sending only logs and metrics can crash the same way.
The victim is just whichever shared-receiver-fed pipeline hasn't been started yet when the server goes live.

**Steps to reproduce**

I wrote a small test — [startorder_race_test.go](https://github.com/tank-500m/opentelemetry-collector/blob/test/sharedreceiver-startorder-race/service/internal/graph/startorder_race_test.go) — that builds a real graph through `graph.Build()` (one `examplereceiver` shared by a traces and a logs pipeline) and then runs `topo.Sort` on the resulting `componentGraph` the same way `StartAll` does.
The graph isn't hand-built; the nodes, per-signal IDs and edges all come from the builder.
It just counts how often the receiver's logs node ends up starting before the traces processor.

```
$ go test -v -count=3 -run TestSharedReceiverStartOrderIsNonDeterministic ./service/internal/graph/
=== RUN   TestSharedReceiverStartOrderIsNonDeterministic
    startorder_race_test.go: iters=5000  panic-window(recvLogs<procTraces)=972   safe=4028
--- PASS
=== RUN   TestSharedReceiverStartOrderIsNonDeterministic
    startorder_race_test.go: iters=5000  panic-window(recvLogs<procTraces)=997   safe=4003
--- PASS
=== RUN   TestSharedReceiverStartOrderIsNonDeterministic
    startorder_race_test.go: iters=5000  panic-window(recvLogs<procTraces)=1031  safe=3969
--- PASS
```

I also reproduced the actual crash end to end.
Built a collector with an `otlp` (gRPC) receiver, k8sattributes and a debug exporter, with the one receiver shared across traces/logs/metrics (separate exporter per signal).
The kube client never syncs (`KUBERNETES_SERVICE_HOST` points at an unreachable address), so `kc` only becomes non-nil once `Start()` runs.
The only code change is a delay at the very top of k8sattributes' `Start()`, gated on an env var so it's a no-op normally.
The panic line itself is untouched; the delay just widens the `kc == nil` window (it's microseconds in practice) enough for in-flight traffic to land in it:

```go
// processor/k8sattributesprocessor/processor.go, first lines of Start()
if d := os.Getenv("K8SATTR_START_DELAY"); d != "" {
    if delay, err := time.ParseDuration(d); err == nil && delay > 0 {
        time.Sleep(delay) // widen the window before kp.kc is set; no-op when unset
    }
}
```

With that I get the panic (gRPC doesn't recover per request, so the process just dies), and the victim signal moves around between logs/metrics/traces run to run:

```
panic: runtime error: invalid memory address or nil pointer dereference
...k8sattributesprocessor.(*kubernetesprocessor).processResource(...)
    processor/k8sattributesprocessor/processor.go:218   // kp.kc.GetPod(...)
...processTraces(...) -> processorhelper -> otlpreceiver Export
google.golang.org/grpc.(*Server).processUnaryRPC(...)
```

For what it's worth, turning on the `ShareProcessorBetweenPipelines` feature gate made the crash go away (0/30 under the same setup): with a single shared processor instance, its `Start()` (and the `kc` assignment) runs before any receiver node brings the server up.
That only works when every fed pipeline uses the same processor id, though.

**What did you expect to see?**

A shared receiver shouldn't start serving a signal until the consumers of every pipeline it feeds are up — including the pipelines for other signals, not just the one whose node happened to trigger the start.

**What did you see instead?**

The receiver's logs node started before the traces pipeline's processor in around 20% of fresh builds (and after it the rest of the time), so the cross-signal order really does go both ways.
Within a single subgraph it's always fine — a receiver never starts before its own processor, and that assertion in the test never trips — so the risk is entirely from the other signal's node, which has no ordering relationship to `proc[traces]` at all.
Add the shared server coming up the instant that first node starts, and that's the `processor.go:202` nil deref.

**Environment**

- collector v0.153.0 / k8sattributesprocessor v1.59.0.
- Related: [contrib #47672][contrib-47672] (the panic), [core #12717][core-12717] (start order).

[contrib-47672]: https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/47672
[core-12717]: https://github.com/open-telemetry/opentelemetry-collector/issues/12717


### Collector version

v0.153.0

### Environment information

## Environment
OS: (e.g., "Ubuntu 20.04")
Compiler(if manually compiled): (e.g., "go 14.2")


### OpenTelemetry Collector configuration

```yaml

```

### Log output

```shell

```

### Additional context

_No response_

### Tip

<sub>[React](https://github.blog/news-insights/product-news/add-reactions-to-pull-requests-issues-and-comments/) with 👍 to help prioritize this issue. Please use comments to provide useful context, avoiding `+1` or `me too`, to help us triage it. Learn more [here](https://opentelemetry.io/community/end-user/issue-participation/).</sub>

--- a/.chloggen/start-receivers-last.yaml
+++ b/.chloggen/start-receivers-last.yaml
@@ -1,0 +1,27 @@
+# Use this changelog template to create an entry for release notes.
+
+# One of 'breaking', 'deprecation', 'new_component', 'enhancement', 'bug_fix'
+change_type: bug_fix
+
+# The name of the component, or a single word describing the area of concern, (e.g. receiver/otlp)
+component: pkg/service
+
+# A brief description of the change.  Surround your text with quotes ("") if it needs to start with a backtick (`).
+note: Ensure receivers always start after all other components
+
+# One or more tracking issues or pull requests related to the change
+issues: [15495]
+
+# (Optional) One or more lines of additional information to render under the primary note.
+# These lines will be padded with 2 spaces and then inserted directly into the document.
+# Use pipe (|) for multiline entries.
+subtext: |
+  There was previously a race condition where multiple receivers using a shared internal implementation,
+  such as the OTLP receiver, could start sending telemetry into a pipeline before all its components had fully started.
+
+# Optional: The change log or logs in which this entry should be included.
+# e.g. '[user]' or '[user, api]'
+# Include 'user' if the change is relevant to end users.
+# Include 'api' if there is a change to a library API.
+# Default: '[user]'
+change_logs: []

--- a/service/internal/graph/graph.go
+++ b/service/internal/graph/graph.go
@@ -407,7 +407,7 @@ func (g *Graph) StartAll(ctx context.Context, host *Host) error {
 	}
 
 	nodes, err := topo.SortStabilized(g.componentGraph, func(nodes []graph.Node) {
-		slices.SortFunc(nodes, func(node1 graph.Node, node2 graph.Node) int {
+		slices.SortFunc(nodes, func(node1, node2 graph.Node) int {
 			// Always start receivers after non-receivers, to avoid cases where a shared receiver
 			// is started from one pipeline and starts sending to another, not yet started pipeline.
 			_, isReceiver1 := node1.(*receiverNode)

--- a/service/internal/graph/graph.go
+++ b/service/internal/graph/graph.go
@@ -13,6 +13,7 @@
 package graph // import "go.opentelemetry.io/collector/service/internal/graph"
 
 import (
+	"cmp"
 	"context"
 	"errors"
 	"fmt"
@@ -405,7 +406,21 @@ func (g *Graph) StartAll(ctx context.Context, host *Host) error {
 		return errors.New("host cannot be nil")
 	}
 
-	nodes, err := topo.Sort(g.componentGraph)
+	nodes, err := topo.SortStabilized(g.componentGraph, func(nodes []graph.Node) {
+		slices.SortFunc(nodes, func(node1 graph.Node, node2 graph.Node) int {
+			// Always start receivers after non-receivers, to avoid cases where a shared receiver
+			// is started from one pipeline and starts sending to another, not yet started pipeline.
+			_, isReceiver1 := node1.(*receiverNode)
+			_, isReceiver2 := node2.(*receiverNode)
+			if isReceiver1 && !isReceiver2 {
+				return -1
+			}
+			if isReceiver2 && !isReceiver1 {
+				return 1
+			}
+			return cmp.Compare(node1.ID(), node2.ID())
+		})
+	})
 	if err != nil {
 		return err
 	}

--- a/service/internal/graph/graph_test.go
+++ b/service/internal/graph/graph_test.go
@@ -2636,34 +2636,34 @@ func (ter *testEagerReceiver) Shutdown(ctx context.Context) error {
 
 var _ receiver.Traces = (*testEagerReceiver)(nil)
 
-// This processor returns an error if it receives traces before it has been started
-type testStartableProcessor struct {
+// This exporter returns an error if it receives traces before it has been started
+type testStartableExporter struct {
 	started bool
 }
 
-func (tsp *testStartableProcessor) Capabilities() consumer.Capabilities {
+func (tse *testStartableExporter) Capabilities() consumer.Capabilities {
 	return consumer.Capabilities{}
 }
-func (tsp *testStartableProcessor) ConsumeTraces(ctx context.Context, td ptrace.Traces) error {
-	if !tsp.started {
+func (tse *testStartableExporter) ConsumeTraces(ctx context.Context, td ptrace.Traces) error {
+	if !tse.started {
 		return errors.New("not started")
 	}
 	return nil
 }
-func (tsp *testStartableProcessor) Start(ctx context.Context, host component.Host) error {
-	tsp.started = true
+func (tse *testStartableExporter) Start(ctx context.Context, host component.Host) error {
+	tse.started = true
 	return nil
 }
-func (tsp *testStartableProcessor) Shutdown(ctx context.Context) error {
+func (tse *testStartableExporter) Shutdown(ctx context.Context) error {
 	return nil
 }
 
-var _ processor.Traces = (*testStartableProcessor)(nil)
+var _ exporter.Traces = (*testStartableExporter)(nil)
 
 func TestSharedReceiverStartOrder(t *testing.T) {
-	// Test that a shared receiver will not start before a processor it sends to
+	// Test that a shared receiver will not start before an exporter it sends to
 
-	// Define component factories:
+	// Define component factories
 
 	sharedType := component.MustNewType("shared")
 	var sharedReceiver *testEagerReceiver
@@ -2677,21 +2677,19 @@ func TestSharedReceiverStartOrder(t *testing.T) {
 	)
 
 	startableType := component.MustNewType("startable")
-	startableProcessorFactory := processor.NewFactory(
+	startableExporterFactory := exporter.NewFactory(
 		startableType,
 		func() component.Config { return nil },
-		processor.WithTraces(func(ctx context.Context, s processor.Settings, c component.Config, t consumer.Traces) (processor.Traces, error) {
-			return &testStartableProcessor{}, nil
+		exporter.WithTraces(func(ctx context.Context, s exporter.Settings, c component.Config) (exporter.Traces, error) {
+			return &testStartableExporter{}, nil
 		}, component.StabilityLevelStable),
 	)
-	nopFactory := exportertest.NewNopFactory()
-	nopType := nopFactory.Type()
 
 	// To check that the outcome isn't affected by how the topological sort decides to arbitrarily order nodes from separate pipelines,
 	// we try 10 different names for one instance of the shared receiver, resulting in 10 different node IDs and 10 different initial node orders.
 	for i := range 10 {
 		sharedReceiver = &testEagerReceiver{} // Reset shared instance
-		otherReceiverId := component.NewIDWithName(sharedType, fmt.Sprintf("%d", i))
+		otherReceiverName := fmt.Sprintf("%d", i)
 
 		// Build the pipeline
 		set := Settings{
@@ -2699,33 +2697,28 @@ func TestSharedReceiverStartOrder(t *testing.T) {
 			BuildInfo: component.NewDefaultBuildInfo(),
 			ReceiverBuilder: builders.NewReceiver(
 				map[component.ID]component.Config{
-					component.NewID(sharedType): sharedReceiverFactory.CreateDefaultConfig(),
-					otherReceiverId:             sharedReceiverFactory.CreateDefaultConfig(),
+					component.NewID(sharedType):                            sharedReceiverFactory.CreateDefaultConfig(),
+					component.NewIDWithName(sharedType, otherReceiverName): sharedReceiverFactory.CreateDefaultConfig(),
 				},
 				map[component.Type]receiver.Factory{sharedType: sharedReceiverFactory},
 			),
-			ProcessorBuilder: builders.NewProcessor(
-				map[component.ID]component.Config{component.NewID(startableType): startableProcessorFactory.CreateDefaultConfig()},
-				map[component.Type]processor.Factory{startableType: startableProcessorFactory},
-			),
+			ProcessorBuilder: builders.NewProcessor(nil, nil),
 			ExporterBuilder: builders.NewExporter(
-				map[component.ID]component.Config{component.NewID(nopType): nopFactory.CreateDefaultConfig()},
-				map[component.Type]exporter.Factory{nopType: nopFactory},
+				map[component.ID]component.Config{
+					component.NewID(startableType):              startableExporterFactory.CreateDefaultConfig(),
+					component.NewIDWithName(startableType, "2"): startableExporterFactory.CreateDefaultConfig(),
+				},
+				map[component.Type]exporter.Factory{startableType: startableExporterFactory},
 			),
-			ConnectorBuilder: builders.NewConnector(
-				map[component.ID]component.Config{},
-				map[component.Type]connector.Factory{},
-			),
+			ConnectorBuilder: builders.NewConnector(nil, nil),
 			PipelineConfigs: pipelines.Config{
 				pipeline.NewID(pipeline.SignalTraces): {
-					Receivers:  []component.ID{component.NewID(sharedType)},
-					Processors: []component.ID{component.NewID(startableType)},
-					Exporters:  []component.ID{component.NewID(nopType)},
+					Receivers: []component.ID{component.NewID(sharedType)},
+					Exporters: []component.ID{component.NewID(startableType)},
 				},
 				pipeline.NewIDWithName(pipeline.SignalTraces, "2"): {
-					Receivers:  []component.ID{otherReceiverId},
-					Processors: []component.ID{component.NewID(startableType)},
-					Exporters:  []component.ID{component.NewID(nopType)},
+					Receivers: []component.ID{component.NewIDWithName(sharedType, otherReceiverName)},
+					Exporters: []component.ID{component.NewIDWithName(startableType, "2")},
 				},
 			},
 		}

--- a/service/internal/graph/graph_test.go
+++ b/service/internal/graph/graph_test.go
@@ -7,6 +7,7 @@ import (
 	"context"
 	"errors"
 	"fmt"
+	"strconv"
 	"strings"
 	"testing"
 
@@ -2622,7 +2623,7 @@ type testEagerReceiver struct {
 	next []consumer.Traces
 }
 
-func (ter *testEagerReceiver) Start(ctx context.Context, host component.Host) error {
+func (ter *testEagerReceiver) Start(ctx context.Context, _ component.Host) error {
 	td := ptrace.NewTraces()
 	var err error
 	for _, consumer := range ter.next {
@@ -2630,7 +2631,8 @@ func (ter *testEagerReceiver) Start(ctx context.Context, host component.Host) er
 	}
 	return err
 }
-func (ter *testEagerReceiver) Shutdown(ctx context.Context) error {
+
+func (ter *testEagerReceiver) Shutdown(context.Context) error {
 	return nil
 }
 
@@ -2644,17 +2646,20 @@ type testStartableExporter struct {
 func (tse *testStartableExporter) Capabilities() consumer.Capabilities {
 	return consumer.Capabilities{}
 }
-func (tse *testStartableExporter) ConsumeTraces(ctx context.Context, td ptrace.Traces) error {
+
+func (tse *testStartableExporter) ConsumeTraces(context.Context, ptrace.Traces) error {
 	if !tse.started {
 		return errors.New("not started")
 	}
 	return nil
 }
-func (tse *testStartableExporter) Start(ctx context.Context, host component.Host) error {
+
+func (tse *testStartableExporter) Start(context.Context, component.Host) error {
 	tse.started = true
 	return nil
 }
-func (tse *testStartableExporter) Shutdown(ctx context.Context) error {
+
+func (tse *testStartableExporter) Shutdown(context.Context) error {
 	return nil
 }
 
@@ -2680,7 +2685,7 @@ func TestSharedReceiverStartOrder(t *testing.T) {
 	startableExporterFactory := exporter.NewFactory(
 		startableType,
 		func() component.Config { return nil },
-		exporter.WithTraces(func(ctx context.Context, s exporter.Settings, c component.Config) (exporter.Traces, error) {
+		exporter.WithTraces(func(context.Context, exporter.Settings, component.Config) (exporter.Traces, error) {
 			return &testStartableExporter{}, nil
 		}, component.StabilityLevelStable),
 	)
@@ -2689,7 +2694,7 @@ func TestSharedReceiverStartOrder(t *testing.T) {
 	// we try 10 different names for one instance of the shared receiver, resulting in 10 different node IDs and 10 different initial node orders.
 	for i := range 10 {
 		sharedReceiver = &testEagerReceiver{} // Reset shared instance
-		otherReceiverName := fmt.Sprintf("%d", i)
+		otherReceiverName := strconv.Itoa(i)
 
 		// Build the pipeline
 		set := Settings{

--- a/service/internal/graph/graph_test.go
+++ b/service/internal/graph/graph_test.go
@@ -2627,7 +2627,7 @@ func (ter *testEagerReceiver) Start(ctx context.Context, _ component.Host) error
 	td := ptrace.NewTraces()
 	var err error
 	for _, consumer := range ter.next {
-		err = errors.Join(consumer.ConsumeTraces(ctx, td))
+		err = errors.Join(err, consumer.ConsumeTraces(ctx, td))
 	}
 	return err
 }

--- a/service/internal/graph/graph_test.go
+++ b/service/internal/graph/graph_test.go
@@ -5,6 +5,7 @@ package graph
 
 import (
 	"context"
+	"errors"
 	"fmt"
 	"strings"
 	"testing"
@@ -21,6 +22,7 @@ import (
 	"go.opentelemetry.io/collector/consumer/xconsumer"
 	"go.opentelemetry.io/collector/exporter"
 	"go.opentelemetry.io/collector/exporter/exportertest"
+	"go.opentelemetry.io/collector/pdata/ptrace"
 	"go.opentelemetry.io/collector/pdata/testdata"
 	"go.opentelemetry.io/collector/pdata/xpdata/pref"
 	"go.opentelemetry.io/collector/pipeline"
@@ -2612,5 +2614,128 @@ func testGraphBuildErrors(t *testing.T) {
 			_, err := Build(context.Background(), set)
 			assert.EqualError(t, err, tt.expected)
 		})
+	}
+}
+
+// This receiver sends a dummy payload to all its consumers as soon as it starts
+type testEagerReceiver struct {
+	next []consumer.Traces
+}
+
+func (ter *testEagerReceiver) Start(ctx context.Context, host component.Host) error {
+	td := ptrace.NewTraces()
+	var err error
+	for _, consumer := range ter.next {
+		err = errors.Join(consumer.ConsumeTraces(ctx, td))
+	}
+	return err
+}
+func (ter *testEagerReceiver) Shutdown(ctx context.Context) error {
+	return nil
+}
+
+var _ receiver.Traces = (*testEagerReceiver)(nil)
+
+// This processor returns an error if it receives traces before it has been started
+type testStartableProcessor struct {
+	started bool
+}
+
+func (tsp *testStartableProcessor) Capabilities() consumer.Capabilities {
+	return consumer.Capabilities{}
+}
+func (tsp *testStartableProcessor) ConsumeTraces(ctx context.Context, td ptrace.Traces) error {
+	if !tsp.started {
+		return errors.New("not started")
+	}
+	return nil
+}
+func (tsp *testStartableProcessor) Start(ctx context.Context, host component.Host) error {
+	tsp.started = true
+	return nil
+}
+func (tsp *testStartableProcessor) Shutdown(ctx context.Context) error {
+	return nil
+}
+
+var _ processor.Traces = (*testStartableProcessor)(nil)
+
+func TestSharedReceiverStartOrder(t *testing.T) {
+	// Test that a shared receiver will not start before a processor it sends to
+
+	// Define component factories:
+
+	sharedType := component.MustNewType("shared")
+	var sharedReceiver *testEagerReceiver
+	sharedReceiverFactory := receiver.NewFactory(
+		sharedType,
+		func() component.Config { return nil },
+		receiver.WithTraces(func(_ context.Context, _ receiver.Settings, _ component.Config, next consumer.Traces) (receiver.Traces, error) {
+			sharedReceiver.next = append(sharedReceiver.next, next)
+			return sharedReceiver, nil
+		}, component.StabilityLevelStable),
+	)
+
+	startableType := component.MustNewType("startable")
+	startableProcessorFactory := processor.NewFactory(
+		startableType,
+		func() component.Config { return nil },
+		processor.WithTraces(func(ctx context.Context, s processor.Settings, c component.Config, t consumer.Traces) (processor.Traces, error) {
+			return &testStartableProcessor{}, nil
+		}, component.StabilityLevelStable),
+	)
+	nopFactory := exportertest.NewNopFactory()
+	nopType := nopFactory.Type()
+
+	// To check that the outcome isn't affected by how the topological sort decides to arbitrarily order nodes from separate pipelines,
+	// we try 10 different names for one instance of the shared receiver, resulting in 10 different node IDs and 10 different initial node orders.
+	for i := range 10 {
+		sharedReceiver = &testEagerReceiver{} // Reset shared instance
+		otherReceiverId := component.NewIDWithName(sharedType, fmt.Sprintf("%d", i))
+
+		// Build the pipeline
+		set := Settings{
+			Telemetry: componenttest.NewNopTelemetrySettings(),
+			BuildInfo: component.NewDefaultBuildInfo(),
+			ReceiverBuilder: builders.NewReceiver(
+				map[component.ID]component.Config{
+					component.NewID(sharedType): sharedReceiverFactory.CreateDefaultConfig(),
+					otherReceiverId:             sharedReceiverFactory.CreateDefaultConfig(),
+				},
+				map[component.Type]receiver.Factory{sharedType: sharedReceiverFactory},
+			),
+			ProcessorBuilder: builders.NewProcessor(
+				map[component.ID]component.Config{component.NewID(startableType): startableProcessorFactory.CreateDefaultConfig()},
+				map[component.Type]processor.Factory{startableType: startableProcessorFactory},
+			),
+			ExporterBuilder: builders.NewExporter(
+				map[component.ID]component.Config{component.NewID(nopType): nopFactory.CreateDefaultConfig()},
+				map[component.Type]exporter.Factory{nopType: nopFactory},
+			),
+			ConnectorBuilder: builders.NewConnector(
+				map[component.ID]component.Config{},
+				map[component.Type]connector.Factory{},
+			),
+			PipelineConfigs: pipelines.Config{
+				pipeline.NewID(pipeline.SignalTraces): {
+					Receivers:  []component.ID{component.NewID(sharedType)},
+					Processors: []component.ID{component.NewID(startableType)},
+					Exporters:  []component.ID{component.NewID(nopType)},
+				},
+				pipeline.NewIDWithName(pipeline.SignalTraces, "2"): {
+					Receivers:  []component.ID{otherReceiverId},
+					Processors: []component.ID{component.NewID(startableType)},
+					Exporters:  []component.ID{component.NewID(nopType)},
+				},
+			},
+		}
+
+		pg, err := Build(context.Background(), set)
+		require.NoError(t, err)
+
+		// Check that no "not started" errors bubble up from the processors during startup
+		require.NoError(t, pg.StartAll(context.Background(), &Host{
+			Reporter: status.NewReporter(func(*componentstatus.InstanceID, *componentstatus.Event) {}, func(error) {}),
+		}))
 	}
 }

